numpy/_core/src/multiarray/stringdtype/sorts.c

```diff
@@ -0,0 +1,303 @@
+
+int
+stringdtype_stablesort(PyArrayMethod_Context *context, char *const *data,
+                       const npy_intp *dimensions, const npy_intp *NPY_UNUSED(strides),
+                       NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_Descr *descr = context->descriptors[0];
+    PyArray_StringDTypeObject *sdescr = (PyArray_StringDTypeObject *)descr;
+    npy_string_allocator *allocator = NpyString_acquire_allocator(sdescr);
+
+    PyArrayMethod_SortParameters *parameters = (PyArrayMethod_SortParameters *)context->parameters;
+    int descending = (parameters->flags & NPY_SORT_DESCENDING);
+
+    int num = (int)dimensions[0];
+    npy_intp elsize = descr->elsize;
+    char *pl = (char *)data[0];
+    char *pr = pl + num * elsize;
+    char *pw;
+    char *vp;
+    int err = -1;
+
+    /* Items that have zero size don't make sense to sort */
+    if (elsize == 0) {
+        return 0;
+    }
+
+    pw = (char *)malloc((num >> 1) * elsize);
+    vp = (char *)malloc(elsize);
+
+    if (pw != NULL && vp != NULL) {
+        string_mergesort0(pl, pr, pw, vp, elsize, sdescr, descending);
+        err = 0;
+    }
+
+    free(vp);
+    free(pw);
+    NpyString_release_allocator(allocator);
+
+    return err;
+}
+
+int
+stringdtype_sort_get_loop(PyArrayMethod_Context *context, int aligned,
+                          int move_references, const npy_intp *strides,
+                          PyArrayMethod_StridedLoop **out_loop,
+                          NpyAuxData **out_transferdata, NPY_ARRAYMETHOD_FLAGS *flags)
+{
+    *flags |= NPY_METH_NO_FLOATINGPOINT_ERRORS;
+    *out_loop = &stringdtype_stablesort;  // default to mergesort
+    return 0;
+}
+
+static void
+string_amergesort0(npy_intp *pl, npy_intp *pr, char *v, npy_intp *pw, npy_intp elsize,
+                   PyArray_StringDTypeObject *sdescr, int descending)
+{
+    char *vp;
+    npy_intp vi, *pi, *pj, *pk, *pm;
+
+    if (pr - pl > SMALL_STRING_MERGESORT) {
+        /* merge sort */
+        pm = pl + ((pr - pl) >> 1);
+        string_amergesort0(pl, pm, v, pw, elsize, sdescr, descending);
+        string_amergesort0(pm, pr, v, pw, elsize, sdescr, descending);
+        for (pi = pw, pj = pl; pj < pm;) {
+            *pi++ = *pj++;
+        }
+        pi = pw + (pm - pl);
+        pj = pw;
+        pk = pl;
+        while (pj < pi && pm < pr) {
+            if (_cmp(v + (*pm) * elsize, v + (*pj) * elsize, sdescr, descending) < 0) {
+                *pk++ = *pm++;
+            }
+            else {
+                *pk++ = *pj++;
+            }
+        }
+        while (pj < pi) {
+            *pk++ = *pj++;
+        }
+    }
+    else {
+        /* insertion sort */
+        for (pi = pl + 1; pi < pr; ++pi) {
+            vi = *pi;
+            vp = v + vi * elsize;
+            pj = pi;
+            pk = pi - 1;
+            while (pj > pl && _cmp(vp, v + (*pk) * elsize, sdescr, descending) < 0) {
+                *pj-- = *pk--;
+            }
+            *pj = vi;
+        }
+    }
+}
```

Comment:
this too - can we maybe rely on the C++ standard library instead of having an algorithm implementation here in the stringdtype internals?